### PR TITLE
[Modules] New module Azure Digital Twins

### PR DESCRIPTION
**Pull request by ryanmstephens**
Created new module for Azure Digital Twins: https://azure.microsoft.com/en-gb/products/digital-twins/#overview

New module doesn't have any dependencies. This module creates Azure Digital Twins Instances and Digital Twins Instances Endpoints. To the following services: Event Hub, Event Grid and Service Bus.

## Pipeline references

| Pipeline |
| - |
|[DigitalTwins: digitaltwinsinstances](https://github.com/ryanmstephens/ResourceModules/actions/workflows/ms.digitaltwins.digitaltwinsinstances.yml) |

# Type of Change

- [x] New feature (non-breaking change which adds functionality)

# Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/deploy.bicep [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+  name: name
+  location: location
+  identity: identity
+  tags: tags
+  properties: {
+    publicNetworkAccess: !empty(publicNetworkAccess) ? any(publicNetworkAccess) : (!empty(privateEndpoints) ? 'Disabled' : 'Enabled')
+  }
+}
+
+resource eventhubendpointid 'Microsoft.DigitalTwins/digitalTwinsInstances/endpoints@2022-05-31' = if ((!empty(eventhubEndpoint) && contains(eventhubEndpoint, 'EventHubURI'))) {
```
Comment (MrMCake) [inline modules/Microsoft.DigitalTwins/digitalTwinsInstances/deploy.bicep:140]: As per our structure this should be its own child-module (in a sub-folder). It does not provide any capabilities to the module itself, but enabels a consumer to deploy an endpoint separatly as the child is automatically published as it's own template reference.

Comment (MrMCake) [reply modules/Microsoft.DigitalTwins/digitalTwinsInstances/deploy.bicep:140]: The VM module with its extensions may be a good reference.

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/deploy.bicep [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+    lock: contains(privateEndpoint, 'lock') ? privateEndpoint.lock : lock
+    privateDnsZoneGroup: contains(privateEndpoint, 'privateDnsZoneGroup') ? privateEndpoint.privateDnsZoneGroup : {}
+    roleAssignments: contains(privateEndpoint, 'roleAssignments') ? privateEndpoint.roleAssignments : []
+    tags: contains(privateEndpoint, 'tags') ? privateEndpoint.tags : {}
+    manualPrivateLinkServiceConnections: contains(privateEndpoint, 'manualPrivateLinkServiceConnections') ? privateEndpoint.manualPrivateLinkServiceConnections : []
+    customDnsConfigs: contains(privateEndpoint, 'customDnsConfigs') ? privateEndpoint.customDnsConfigs : []
+  }
+}]
+
+resource server_lock 'Microsoft.Authorization/locks@2017-04-01' = if (!empty(lock)) {
```
Comment (MrMCake) [inline modules/Microsoft.DigitalTwins/digitalTwinsInstances/deploy.bicep:211]:
```suggestion
resource digitaltwin_lock 'Microsoft.Authorization/locks@2017-04-01' = if (!empty(lock)) {
```

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/deploy.bicep [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+    condition: contains(roleAssignment, 'condition') ? roleAssignment.condition : ''
+    delegatedManagedIdentityResourceId: contains(roleAssignment, 'delegatedManagedIdentityResourceId') ? roleAssignment.delegatedManagedIdentityResourceId : ''
+    resourceId: digitaltwin.id
+  }
+}]
+
+@description('The resource ID of the Digital Twin.')
+output resourceId string = digitaltwin.id
+
+@description('The name of the resource group the key vault was created in.')
```
Comment (MrMCake) [inline modules/Microsoft.DigitalTwins/digitalTwinsInstances/deploy.bicep:249]:
```suggestion
@description('The name of the resource group the key vault was created in.')
```
```suggestion
@description('The name of the resource group the Digital Twin was created in.')
```

**Comment by MrMCake** [comment]
Hey @ryanmstephens ,

thank you for your contribution. Great work - aside form 2 small suggestions, the only ask would be to move the `endpoint` child-resource into it's own child-resource folder (sub-folder). In this case, the deployments of the resource would just change to deployments of set child-module.

**Comment by ryanmstephens** [comment]
Hello @MrMCake,

Thank you for taking the time to review. 

The above was my initial approach. However, I ran into an issue when moving the endpoints to a child module template: 

`New-AzResourceGroupDeployment:- Error: Code=MissingArgument; Message=Target - 'Test/Test'. Input parameter 'source' is missing`

 Source doesn't seem to be an input parameter for the API:   Microsoft.DigitalTwins/digitalTwinsInstances/endpoints@2022-05-31 

Any thoughts on this? I am looking into raising the issue with PG. Same issue occurs on older API versions as well.

**Comment by AlexanderSehr** [comment]
> Hello @MrMCake,
> 
> Thank you for taking the time to review.
> 
> The above was my initial approach. However, I ran into an issue when moving the endpoints to a child module template:
> 
> `New-AzResourceGroupDeployment:- Error: Code=MissingArgument; Message=Target - 'Test/Test'. Input parameter 'source' is missing`
> 
> Source doesn't seem to be an input parameter for the API: Microsoft.DigitalTwins/digitalTwinsInstances/endpoints@2022-05-31
> 
> Any thoughts on this? I am looking into raising the issue with PG. Same issue occurs on older API versions as well.

Hey @ryanmstephens I'm terribly sorry the late reply - I've been underwater and afraid nobody else picked it up as I commented first, so that's on me. 

That being said - to your question I'd be curious to see the implementation. What I'd expect to see is something like this:

`deploy.bicep`
```bicep
module endpoint 'endpoints/deploy.bicep' = if ((!empty(eventhubEndpoint) && contains(eventhubEndpoint, 'EventHubURI'))) {
  name: '${uniqueString(deployment().name, location)}-digitalTwin-endpoint'
  params: {
    digitalTwinsInstanceName: digitaltwin.name
    name: eventhubEndpoint.name
    (..)
  }
}
```

unless there can be more than one endpoint - which would mean it should have an array syntax. The KeyVault and its accessPolicies / secrets are a good reference.

**Comment by ChrisSidebotham** [comment]
@AlexanderSehr - I actually looked at this with Ryan during the development phases and it appears that `source` parameters is not listed from the resource provider as a valid input parameter when passing as a child module, however the provider requires in. 

When using in the file as the DT itself the source param is not required and seems to be fulfilled by a dynamic value from the parent/child relationship. Very odd experience and I believe @ryanmstephens raised this with PG and did not get a response.

**Comment by AlexanderSehr** [comment]
Hey @ryanmstephens & @ChrisSidebotham , this PR has been running a lot longer than it should - also because I always wanted to give it a spin on my end and didn't find time.
That being said, I'll go ahead an merge this PR into branch in CARML so that your @ryanmstephens is in, and will then open a PR to main after messing with it a bit myself.

**Pull request merged**